src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,154 @@
+        /// </summary>
+        private long _nextResultId;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public DocumentPullDiagnosticHandler(
+            ILspSolutionProvider solutionProvider,
+            IDiagnosticService diagnosticService) : base(solutionProvider)
+        {
+            _diagnosticService = diagnosticService;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+        }
+
+        private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateArgs)
+        {
+            if (updateArgs.DocumentId == null)
+                return;
+
+            _documentIdToLastResultId.Remove((updateArgs.Workspace, updateArgs.DocumentId));
+        }
+
+        public override async Task<DiagnosticReport[]?> HandleRequestAsync(
+            DocumentDiagnosticsParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(request.TextDocument, $"Got a DocumentPullDiagnostic request that did not specify a {nameof(request.TextDocument)}");
+
+            var document = SolutionProvider.GetDocument(request.TextDocument, context.ClientName);
+            if (document == null)
+            {
+                // Client is asking server about a document that no longer exists (i.e. was removed/deleted from the
+                // workspace).  In that case we need to return an actual diagnostic report with `null` for the
+                // diagnostics to let the client know to dump that file entirely.
+                return new[] { new DiagnosticReport { ResultId = null, Diagnostics = null } };
+            }
+
+            var project = document.Project;
+            var solution = project.Solution;
+            var workspace = solution.Workspace;
+
+            // If the client has already asked for diagnostics for this document, see if we have actually recorded any
+            // differences, or if they should just use the same diagnostics as before.
+            if (request.PreviousResultId != null)
+            {
+                lock (_gate)
+                {
+                    if (_documentIdToLastResultId.TryGetValue((workspace, document.Id), out var lastReportedResultId) &&
+                        lastReportedResultId == request.PreviousResultId)
+                    {
+                        // Nothing changed between the last request and this one.  Report a null response to the client
+                        // to know they don't need to do anything.
+                        return null;
+                    }
+                }
+            }
+
+            // Being asked about this document for the first time.  Or being asked again and we have different diagnostics.
+
+            var diagnosticKeys = _diagnosticService.GetDiagnosticsUpdatedEventArgs(workspace, project.Id, document.Id, cancellationToken).ToImmutableArray();
```

Comment:
How does GetDiagnosticsUpdatedEventArgs work exactly?  Presumably it gets whatever the updates are from the last diagnostic service update, but is the last update going to correspond to everything since LSP asked us?

Why can't we just simply call GetDiagnostics with the project context LSP handed to us?

Reply:
> How does GetDiagnosticsUpdatedEventArgs work exactly?

You can think of the IDiagnosticsService as a mapping from special "keys" to the set of diagnostics for those keys.  This method name is actually *terrible* as these aren't really events or args.  They're really just saying "what are all the keys we know about for this particular set of args".  In this case, we're asking for all the keys associated with a particular document.  We can then use those keys to actually find out all the diagnostics for that doc.

You could imagine this exposed behind a helper method (which is what i think i will do :)).